src/Network/Wai/Log/Internal.hs

```diff
@@ -5,26 +5,30 @@ import Data.Aeson.Types (Value(..), object)
 import Data.ByteString.Builder (Builder)
 import Data.Text (Text)
 import Data.Time.Clock (UTCTime, diffUTCTime, getCurrentTime)
-import Data.UUID (UUID)
+import Data.UUID (toText)
 import Data.UUID.V4 (nextRandom)
 import Log (LogLevel)
-import Network.Wai (Application, responseToStream)
+import Network.Wai (Application, requestHeaders, responseToStream)
+import qualified Data.Text.Encoding as Text
 
-import Network.Wai.Log.Options (Options(..), ResponseTime(..), requestUUID)
+import Network.Wai.Log.Options (Options(..), ResponseTime(..), requestID)
 
 -- | This type matches the one returned by 'getLoggerIO'
 type LoggerIO = UTCTime -> LogLevel -> Text -> Value -> IO ()
 
 -- | Create a logging 'Middleware' that takes request UUID
 -- given a 'LoggerIO' logging function and 'Options'
-logRequestsWith :: LoggerIO -> Options -> (UUID -> Application) -> Application
+logRequestsWith :: LoggerIO -> Options -> (Text -> Application) -> Application
 logRequestsWith loggerIO Options{..} mkApp req respond = do
-  uuid <- nextRandom
-  logIO "Request received" $ logRequest uuid req
+  -- If the `request_id` can't be found or parsed successfully from the header generated one instead
+  reqID <- case lookup "X-Request-ID" (requestHeaders req) of
+    Nothing -> toText <$> nextRandom
```

Comment:
What should we be using to generate `request_id`s instead of `nextRandom` to emulate the 32-bit hexadecimal `id`s generated by nginx?

Reply:
I don't believe it's important to emulate the shape of Nginx's request ID because the whole idea is that `wai-log` handles an opaque ID. I would even go as far as maintaining a minimal amount of invariants (implicit or explicit), and the 32-bit hex part is not one that I'd suggest we support. It must be of reasonable length and with printable characters that are easy for humans to distinguish, and I don't find anything more to enforce or mimic from outside.

Reply:
I also think the ID should be opaque.  We could go as far as using a type variable, I think the only requirement from the library's implementation point of view is that the ID has `ToJSON` instance (haven't checked though).